### PR TITLE
Final

### PR DESCRIPTION
**Pull request by KayanatSuleman**

### CODE REVIEW

File: Section_1_Theory_Answers.md [hunk 1]
```diff
@@ -126,4 +126,6 @@ ZeroDivisionError: division by zero
 
 + We can integrate API to help with automation of certain tasks
 
-+ And finally to create a secure environment and ensure secure login and data encryption we could use a tool such as OAuth.
\ No newline at end of file
++ And finally to create a secure environment and ensure secure login and data encryption we could use a tool such as OAuth.
+
+done
```
Comment (Omotola28) [inline Section_1_Theory_Answers.md:131]: 31/31 => Good work

File: Section_2_Answer.py [hunk 2]
```diff
@@ -15,4 +15,6 @@ def is_isogram(str_input):
             return False
         unique_letters.add(lowercase_letter)
 
-    return True
\ No newline at end of file
+    return True
+
+    done
```
Comment (Omotola28) [inline Section_2_Answer.py:20]: 7/7

File: Section_2_Test_File.py [hunk 3]
```diff
@@ -19,3 +19,5 @@ def test_is_isogram():
     print("All test cases passed")
 
 test_is_isogram()
+
+done
```
Comment (Omotola28) [inline Section_2_Test_File.py:23]: 8/12 => Your test file is not well structured you should have had something like this

```
import unittest
from Section_2 import isogram_checker
# importing the unittest module and the function we are testing

class TestingIsogramFunction(unittest.TestCase):

    def test_isogram_checker_true(self):
        expected = True
        result = isogram_checker("pie")
        self.assertEqual(expected, result)
    # checks that an isogram gets a true result using an example of "pie" as our input word

    def test_isogram_checker_false(self):
        expected = False
        result = isogram_checker("green")
        self.assertEqual(expected, result)
    # checks that a word that is not an isogram gets a false result using an example of "green"

    def test_isogram_checker_int(self):
        with self.assertRaises(TypeError):
            isogram_checker(4)
    # Tests than an invalid input (in this case if you use a int instead of a string) raises a type error

    def test_isogram_checker_edge_case_same_char(self):
        expected = False
        result = isogram_checker("eeeeeeeee")
        self.assertEqual(expected, result)
    # Tests for an edge case where instead of inputting a word, we just get a collection of the same letter - should return false

    def test_isogram_checker_edge_case_spaces(self):
        expected = True
        result = isogram_checker("pie mouth")
        self.assertEqual(expected, result)
    # Tests for an edge case where two words are used in the input
    # If i had more time I would expand the isogram function to split out words in a string and apply the function to each word in the string
```

File: Section_3_Answer.py [hunk 4]
```diff
@@ -42,3 +42,4 @@ def allocate_classes(total_students):
     print(allocation)
     return allocation
 
+done
```
Comment (Omotola28) [inline Section_3_Answer.py:45]: 25/25 => Really good work

File: Section_4_Answer.py [hunk 5]
```diff
@@ -1,3 +1,5 @@
 '''
 Please refer to Kayanat_Suleman_Main_Assessment_2_Answers.pdf in Slack.
-'''
\ No newline at end of file
+'''
```
Comment (Omotola28) [inline Section_4_Answer.py:3]: 25/25

**Review by Omotola28 — APPROVED** [review]
96% => Really great work on the assessment would have been better if the test you wrote was better structured and used the appropriate modules
